btree/btree.c

```diff
@@ -2112,21 +2126,22 @@ static int64_t node_free(struct node_op *op, struct nd *node,
 	int shift = node->n_type->nt_shift(node);
 
 	m0_rwlock_write_lock(&list_lock);
+	node_lock(node);
 	node->n_ref--;
 	node->n_be_node_valid = false;
+	op->no_addr = node->n_addr;
+	m0_free_aligned(segaddr_addr(&op->no_addr), 1ULL << shift, shift);
 
-	if (node->n_ref == 0) {
+	if (node->n_ref == 0 && node->n_txref == 0) {
```

Comment:
What if the node_free and/or node_put gets called before transaction is committed and callback is called? In that case n_ref could reach zero but txref will still be non-zero in the node_free/node_put calls. Similar check should be added in callback to free up node descriptor in callback.

Reply:
For any node which we want to free we are calling btree_tx_nodes_capture for that node and then node_free, so this scenario will not occur. If you see we are calling P_CAPTURE before P_FREENODE. so n_ref will not become zero.

Reply:
In btree_tx_nodes_capture we are just setting the callback btree_tx_commit_cb(). This will be invoked on tx_done, which we do not know when will it happen.

Reply:
@RadhaGulhane13 this needs to be taken care of in the capture callback since we cannot free the descriptor now because even if the n_ref is zero the n_txref is still non-zero. After sometime when the Tx commit completes and the Tx callback gets called then the node's n_txref will be decremented and become zero. At that point we need to free the node descriptor.
@NikhilBirgade  this might need to be done in the node_tx_callback routine !!!

Reply:
So, we have node->n_txref == 0 check as well, and it is getting decremented on in btree_tx_commit_cb()

Reply:
made changes according to discussion.